File: leetcode/linked_list/369_Plus_One_Linked_List/codes/ll.c
struct ListNode {
    int val;
    struct ListNode* next;
};

int helper(struct ListNode* node) {
    if (!node) return 1;
    int carry = helper(node -> next);
    node -> val += carry;
    if (node -> val == 10) {
        node -> val = 0;
        return 1;
    }
    return 0;
}

struct ListNode* plusOne(struct ListNode* head) {
    if (helper(head)) {
        struct ListNode* newHead = malloc(sizeof(struct ListNode));
        newHead -> val = 1;
        newHead -> next = head;
        return newHead;
    }
    return head;
}

File: leetcode/linked_list/369_Plus_One_Linked_List/codes/ll.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def plusOne(self, head: ListNode) -> ListNode:
        def dfs(node: ListNode) -> int:
            if not node:
                return 1 
            carry = dfs(node.next)
            node.val += carry
            if node.val == 10:
                node.val = 0
                return 1
            return 0

        carry = dfs(head)
        if carry == 1:
            new_head = ListNode(1)
            new_head.next = head
            return new_head
        return head
